[Python] Implement new Concept Exercise: none #1031 (#2283)
exercism/v3: languages/python/exercises/concept/none/.docs/introduction.md

When you are beginning to code, you always assign a variable to a value. Say,

```python
a = 3
b = True
```

This declares the variable as well as defines them. But when you want to just declare the variable and not define them at start you use a special built in called as `None`.

> 

Reviewer: For the `introduction.md`, the goal is to focus only on the information needed to explain/define the concept and complete the exercise.  It should be as simple & direct as possible, with example usage.  So I think we can remove these paragraphs about variable declaration, and then go into more detail or nuance in the `after.md`.

Author: Noted. I will refer the docs from other Sources you have given me and alter the introduction.md. 

Although, I should reference them in the design.md ( as well as after.md ) right?

Reviewer: **Correct!**  we definitely **do** want those additional resources available to the student in places like `hints.md` and `after.md`.  And we also want them in the `design.md`, so that we internally have  a record of what resources we used to build the exercise (_and what we might need to alter or update at a later time_), and what our thinking was around the concept/topic.

If you have heard about `NULL` type in other languages then this must be familiar. `None` helps you to declare variables. These variables can be used later to be defined to a value.

```python
a = None
print(a)
#=> None
type(a)
#=> <class 'NoneType'>
```

In Python, `None` is a Singleton built-in which is used to declare a variable as well as can be used as `False`. Meaning if you want to check for a variable which is declared with `None`, It will be considered as `False`

> 

Reviewer: I think it's good here to emphasize the "placeholder"/null nature of `None` -- that it represents an _**absence of value**_,  and also to mention that Python uses `None` when a return from a function is not specified.    But referring to variable declaration (_None can be a placeholder when you don't have a default value, but it really doesn't have a function beyond that_) confuses things here.  -- see [python docs](), [https://docs.python.org/3/library/constants.html#None], [David Beazly's Python Course](https://dabeaz-course.github.io/practical-python/Notes/02_Working_with_data/01_Datatypes.html) (_which mentions the conditional evaluation_), and  [educative.io None keyword](https://www.educative.io/edpresso/what-is-the-none-keyword-in-python) for some wording examples that are a little more direct.

 I also think it's dangerous to tell the student that `None` can be _**used as**_ `False` as a main case.  `None` _**evaluates**_ to `False` when used in a conditional check (_which is something we should cover in this exercise_), but `None` is not of type `bool`, and won't behave as such.  For cases where a `False` type is needed, `bool` type `False` should be used.  "Falsey" and "Truthy" are a different concept exercise.

Reviewer: I also think we should save the discussion about `None` being a Singleton object for the `after.md` (_maybe link to the educative.io page, or some other good explanation??_).  It is very valuable information, but not really necessary for understanding basic usage or for completing this exercise.

Author: Got it. I will mode the discussion of None being a singleton object to after.md 

It is something very necessary to know but also a little tough to understand. I will refer some docs to bring it up more easier.

Reviewer: Agree.  We might want to consider a follow-on concept exercise around `None` being a singleton, or its possible uses as a `sentinel value` -- but it feels like that might be under an OOP category, or some other more "advanced" scenario.  

But speaking to those uses in the `after.md` is a great start -- you can give examples, point the student to some articles or problems as "next steps", or link to projects or code where `None` is used in that fashion.  There are a lot of possibilities for what a student could explore after this basic run-through....we just don't want to chase them away with too much info at the very start.  😉

> In Python, `None` is frequently used to represent the absence of a value, it can also be used to define a `null` variable or an object.

Reviewer: ```suggestion
In Python, `None` is frequently used to represent the absence of a value --  a placeholder to define a `null` (empty) variable, object, or argument.
```

> If you have heard about `NULL` type in other languages then this must be familiar. `None` helps you to declare variables. These variables can be used later to be defined to a value. We can assign `None` to any variable.

Reviewer: ```suggestion
If you've heard about or used the `NULL` or `nil` type in other languages,  then this usage of `None` in Python will be familiar to you. `None` helps you to declare variables that you don't yet have values for. These variables can then be re-assigned to a value later as needed. We can assign `None` to any variable, argument, or attribute.
```

> If you've heard about or used a `NULL` or `nil` type in another programming language, then this usage of `None` in Python will be familiar to you. `None` helps you to declare variables that you don't yet have values for. These variables can then be re-assigned to a value later as needed.

Reviewer: ```suggestion
If you've heard about or used a `NULL` or `nil` type in another programming language, then this usage of `None` in Python will be familiar to you. `None` helps you to declare variables or function arguments that you don't yet have values for. These can then be re-assigned to specific values later as needed.
```

```python
a = None
print(a)
#=> None
type(a)
#=> <class 'NoneType'>
```

> 

Reviewer: So I think it might be goo to add an example here of `none` as a default value for an argument.  
Something like:

```python
def add_to_todos(new_task, todo_list=None):
  if todo_list is None:
    todo_list = []
    todo_list.append(new_task)
  return todo_list
```

> Python will also return `None` from a function that doesn't already have a stated `return` value, making it easy to check if the function completed.

Reviewer: ```suggestion
Python will also return `None` from a function that doesn't already have a stated `return` value, making it easy to check if the function completed without errors.
```

Reviewer: Suggest a quick code example here like:

```python
def no_values_returned():
    pass

>>> print(no_values_returned())
None
```

```python
def test_func():
    pass
test_func()
#=> returns None

def test_func2():
    return
#=> returns None
```

> `None` evaluates to `False` when used in a conditional check.

Reviewer: ```suggestion
`None` will evaluate to `False` when used in a conditional check, so it is useful for validating the "presence of" or "absence of" a value - _any_ value -- a pattern frequently used when a function or process might hand back an error object or message.  
```
